btf/types.go

```diff
@@ -67,13 +71,73 @@ func (ts types) ByID(id TypeID) (Type, error) {
 	return ts[id], nil
 }
 
+type Skipper func(Type) bool
+
+type Comparer func(Type, Type) bool
+
+// SkipQualifierAndTypedef skips qualifiers and Typedefs.
+var SkipQualifierAndTypedef = func(t Type) bool {
+	switch t.(type) {
+	case qualifier, *Typedef:
+		return true
+	}
+	return false
+}
+
+// CompareTypeName returns true if a and b's names correspond.
+var CompareTypeName = func(a, b Type) bool {
+	return a.TypeName() == b.TypeName()
+}
+
+// Equal compares Type a to Type b. By default, only binary equivalency is
+// checked, ignoring type names.
+//
+// cmp allows providing extra comparison rules. It will be evaluated after
+// checking a and b's binary equivalency at each step of the type hierarchy.
+// skip allows individual Types to be omitted from comparisons, and is called
+// before checking a and b's binary equivalency.
+func Equal(a, b Type, cmp Comparer, skip Skipper) error {
+	if a == nil && b == nil {
+		return nil
+	}
+	if a == nil || b == nil {
+		return errors.New("one of a or b are nil")
+	}
+
+	at, bt := postorderTraversal(a, nil), postorderTraversal(b, nil)
```

Comment:
I think pre-order traversal is better here: https://en.wikipedia.org/wiki/Tree_traversal#Data_structures_for_tree_traversal It would mean we fail faster when there is a difference. The implementation of preorder can also be faster since it doesn't need as much temporary storage.

Reply:
Oh absolutely, postorder is much too expensive for this. I initially tried comparing all vmlinux' types against themselves, and that didn't finish within 5 minutes or so. Preorder doesn't require any allocs except for the iterator that has some scratch space, and as you mentioned, would allow us to bail out much sooner.